### PR TITLE
JSON reader improvements/fixes

### PR DESCRIPTION
**Pull request by lnkuiper**
This PR improves the `read_json` functionality, fixes bugs, and changes the API (slightly).

## API Changes
Whereas we first had the parameters `lines` and `json_format`, we now have `records` and `format`. This allows for slightly more kinds of JSON to be read.

Parameter `records` can be `['auto', 'true', 'false']`.
Parameter `format` can be any of `['auto', 'unstructured', 'newline_delimited', 'array']`.

Combinations of these allow most kinds of JSON to be read.

### Records
If the elements in the JSON you are trying to read are "records", i.e., JSON objects where the key/values should be unpacked to columns, then `records` should be set to `true`. However, if you prefer to read the JSON objects as DuckDB `STRUCT`s, then `records` can be set to `false` instead.

If the elements are non-records, e.g., JSON arrays, strings, etc., then `records` should be set to `false`.

In general, however, this parameter be auto-detected.

### Format
If these elements are separated by a newline, i.e., NDJSON, then `format` should be set to `'newline_delimited'` (or `'nd'` in short, or use `read_ndjson` instead).

If they are contained in a JSON array, then `format` should be set to `'array'`. Support for reading JSON arrays has improved greatly, as we can now do streaming reads of these, i.e., we do not require the whole array to fit in a buffer, just individual array elements.

If your elements have no real structure, for example, each JSON file contains one or more pretty-printed (containing newlines) JSON object/array/etc., then `format` should be set to `'unstructured'`.

This parameter can also be auto-detected; as a bonus, this can differ across files. For example:
```sql
select * from read_json_auto(['newline_delimited_objects.ndjson', 'array_of_objects.json']);
```
Should work fine, as long as the objects have the same schema.

## MultiFileReader Integration
This PR also integrates the JSON reader with the `MultiFileReader` introduced in #6912, which adds the `filename`, `union_by_name`, and `hive_partitioning` parameters.

So, if you have different schemas and different formats, you can still read them just fine with:
```sql
select * from read_json_auto(['newline_delimited_objects.ndjson', 'array_of_objects.json'], union_by_name=true);
```
As this will force DuckDB to sample data from each of the files.

We now also support hive-partitioned JSON reads but not yet writes.

## COPY
We now also support creating JSON array files with the `COPY` statement (we only supported NDJSON before).

For example:
```sql
copy (select * from range(5)) to 'my.json' (ARRAY TRUE);
```
Will create the following file:
```json
[
	{"range":0},
	{"range":1},
	{"range":2},
	{"range":3},
	{"range":4}
]
```

Which can be read like so:
```sql
create table test (range bigint);
copy test from 'my.json' (ARRAY TRUE);
copy test from 'my.json' (AUTO_DETECT TRUE);
```

### Bugfixes
This PR also fixes bugs that have accumulated since the JSON reader was released. I've also improved a ton of error messages all over the place, as well as increased the `sample_size` and `maximum_object_size` parameter defaults, to allow for JSON to be read easily without setting parameters.

This PR fixes the following issues:
1. Default to `JSON` type in auto-detection when we see only `NULL` #7448
2. JSON `null` now always goes to our `NULL`, even though `null` is valid JSON (follows Postgres behavior) #6779
3. When forcing columns (e.g., `columns={"ts": "TIMESTAMP[]"}`), we can cast to timestamp without a specifier #6774
4. Reader improvements in this PR (+ cast improvements in PR #7366) fix the following list of issues:
    * #6588
    * #6646
    * #6768
    * #6803 
    * #7015
    * #7103 
    * #7145
    * #7155
    * #7307

### CODE REVIEW

File: test/sql/json/test_json_copy.test_slow [hunk 1]
```diff
@@ -99,12 +99,38 @@ copy conclusions from 'data/json/top_level_array.json'
 statement ok
 copy conclusions from 'data/json/top_level_array.json' (AUTO_DETECT TRUE)
 
+statement ok
+delete from conclusions;
+
+# and also if we say it's an array
+statement ok
+copy conclusions from 'data/json/top_level_array.json' (ARRAY TRUE)
```
Comment (Mytherin) [inline test/sql/json/test_json_copy.test_slow:107]: As mentioned above can we add a test with `ARRAY` without the `TRUE`?

File: test/sql/json/test_json_extract.test [hunk 2]
```diff
@@ -7,6 +7,12 @@ require json
 statement ok
 pragma enable_verification
 
+# should go to our NULL
+query T
+select json_extract('{"foo": null}', '$.foo')
```
Comment (Mytherin) [inline test/sql/json/test_json_extract.test:12]: Can we add a test where we do this with null values in a list, and with only a `null` constant? What about a JSON path expression where a value early on in the path is NULL?

File: test/sql/json/tpch_round_trip.test_slow [hunk 3]
```diff
@@ -2,23 +2,24 @@
...
 
+statement ok
+start transaction;
+
 statement ok
 call dbgen(sf=1)
 
 statement ok
-copy (select to_json(lineitem) from lineitem) to '__TEST_DIR__/lineitem.json' (format csv, quote '')
+copy lineitem to '__TEST_DIR__/lineitem.json'
```
Comment (Mytherin) [inline test/sql/json/tpch_round_trip.test_slow:16]: Can we read/write lineitem as an array-of-records as well, and verify that we are doing a streaming read by using `SET memory_limit=X`?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks great! Some comments below:

File: tools/pythonpkg/tests/extensions/json/data/example.json [hunk 4]
```diff
@@ -1,5 +1,7 @@
-{"id":1,"name":"O Brother, Where Art Thou?"}
-{"id":2,"name":"Home for the Holidays"}
-{"id":3,"name":"The Firm"}
-{"id":4,"name":"Broadcast News"}
-{"id":5,"name":"Raising Arizona"}
\ No newline at end of file
+[
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply tools/pythonpkg/tests/extensions/json/data/example.json:1]: Yes, the default for `read_json` is now `array`, and the default for `read_ndjson` is `newline_delimited`. I did not want to change the python tests as they do not expose the `format` parameter (yet), so I changed the test file.

[the remark this answers is not in the document]
Comment (Mytherin) [reply tools/pythonpkg/tests/extensions/json/data/example.json:1]: Maybe @Tishj can work on adding support for the new parameters to `read_json` and `write_json` in the Python client (possibly with a branch off of this PR)?

File: extension/json/json_scan.cpp [hunk 5]
```diff
@@ -319,7 +331,71 @@ static inline const char *PreviousNewline(const char *ptr) {
...
+						ptr++; // Skip the escaped char
+					}
+				}
+			}
+			break;
+		default:
+			continue;
+		}
+
+		if (parents == 0) {
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply extension/json/json_scan.cpp:358]: Yes, the whitespace between array elements is always skipped over in `JSONScanLocalState::ParseNextChunk`, and `JSONScanLocalState::SkipOverArrayStart`

File: test/sql/json/json_multi_file_reader.test [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+2	Home for the Holidays	NULL
+3	The Firm	NULL
+4	Broadcast News	NULL
+5	Raising Arizona	NULL
+NULL	NULL	cancelled
+NULL	NULL	cancelled
+
+# despite not being able to do partitioned writes, we can do partitioned json reads already!
+query II
+SELECT j, count(*) FROM read_json_auto('__TEST_DIR__/json_part/j=*/*.csv', HIVE_PARTITIONING=1) group by j order by j;
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply test/sql/json/json_multi_file_reader.test:48]: Good idea, added

File: test/sql/json/test_json_copy.test_slow (hunk 1 above)
Comment (lnkuiper) [reply test/sql/json/test_json_copy.test_slow:156]: Good catch, this is now fixed and works as expected.

File: test/sql/json/test_json_copy.test_slow [hunk 7]
```diff
@@ -99,12 +99,38 @@ copy conclusions from 'data/json/top_level_array.json'
...
+
 query I
 select * from conclusions
 ----
 cancelled
 cancelled
 
+# we can also write JSON arrays instead of newline-delimited
+statement ok
+copy (select range as i from range(10)) to '__TEST_DIR__/my.json' (ARRAY TRUE)
```
[the remark this answers is not in the document]
Comment (lnkuiper) [reply test/sql/json/test_json_copy.test_slow:172]: Done. This caught a bug where I did not get the validity by reference but accidentally by value in the map conversion. All types roundtrip now as ND/ARRAY

File: test/sql/json/test_json_extract.test (hunk 2 above)
Comment (lnkuiper) [reply test/sql/json/test_json_extract.test:12]: Done, works as expected

File: test/sql/json/tpch_round_trip.test_slow (hunk 3 above)
Comment (lnkuiper) [reply test/sql/json/tpch_round_trip.test_slow:16]: I already tested reading/writing lineitem as ARRAY in `test_json_copy.test_slow`, added a memory limit there, works as expected.

**Comment by lnkuiper** [comment]
Thanks for the prompt PR feedback @Mytherin! This exposed two bugs, which have now been fixed.

Fingers crossed for the CI 🤞

File: extension/json/json_functions/copy_json.cpp [hunk 8]
```diff
@@ -116,7 +116,9 @@ static unique_ptr<FunctionData> CopyFromJSONBind(ClientContext &context, CopyInf
 		} else if (loption == "compression") {
 			bind_data->SetCompression(StringValue::Get(kv.second.back()));
 		} else if (loption == "array") {
-			bind_data->options.format = JSONFormat::ARRAY;
+			if (BooleanValue::Get(kv.second.back().DefaultCastAs(LogicalTypeId::BOOLEAN))) {
```
Comment (Mytherin) [inline extension/json/json_functions/copy_json.cpp:119]: This still seems incorrect to me (and actually adds the problem in a different location). For example:

```sql
D CREATE TABLE integers(i INT);
D COPY integers to 'integers.json' (ARRAY);
ABORT THROWN BY INTERNAL EXCEPTION: 'back' called on an empty vector!
zsh: abort      build/debug/duckdb
```

The problem is that when only `ARRAY` is specified (without `TRUE` or `FALSE`) `kv.second` is empty as no parameters are provided. That is generally interpreted as `TRUE` in other copy statements (e.g. for the CSV reader `HEADER` is equivalent to `HEADER TRUE`).

Comment (lnkuiper) [reply extension/json/json_functions/copy_json.cpp:119]: Ah now I understand how this works. Thanks! Working on a fix

Comment (lnkuiper) [reply extension/json/json_functions/copy_json.cpp]: Got it. Properly checking the parameters now and throwing errors when necessary.

**Comment by Mytherin** [comment]
Had a go at fixing the CI failures and added some more tests for JSON writing. The CI failure was relatively minor (two newlines being written instead of 1 for empty CSV files) and the new tests I added all passed, with one exception - the following empty file can no longer be parsed:

```json
[
	
]
```
This results in the following error:
```
Error: Invalid Input Error: Malformed JSON in file "empty.json", at byte 1 in record/value 2: unexpected character. 
```

v0.7.1 parses it correctly:

```
duckdb -c "FROM 'empty.json'"
┌────────┐
│  json  │
│ int32  │
├────────┤
│ 0 rows │
└────────┘
```

**Pull request merged**